ts/src/header-validator/validate-json.ts

```diff
@@ -1259,40 +1265,107 @@ function aggregatableValue(ctx: Context, j: Json): Maybe<number> {
 
 function aggregatableKeyValue(
   ctx: Context,
-  [key, j]: [string, Json]
-): Maybe<number> {
+  [key, j]: [string, Json],
+  maxBytes: Maybe<number>
+): Maybe<AggregatableValuesValue> {
   if (!aggregationKeyIdentifierLength(ctx, key, 'key ')) {
     return None
   }
-  return aggregatableValue(ctx, j)
+
+  return typeSwitch(ctx, j, {
+    number: (ctx, j) =>
+      aggregatableKeyValueValue(ctx, j).map((j) => ({
+        value: j,
+        filteringId: constants.defaultFilteringIdValue,
+      })),
+    object: (ctx, j) =>
+      struct(ctx, j, {
+        value: field('value', aggregatableKeyValueValue),
+        filteringId: field('filtering_id', (ctx, j) =>
+          uint(ctx, j).filter((n) => {
+            if (maxBytes.value === undefined) {
+              ctx.error(
+                `cannot be fully validated without a valid aggregatable_filtering_id_max_bytes`
+              )
+              return false
+            }
+            return isInRange(
+              ctx,
+              n,
+              0n,
+              256n ** BigInt(maxBytes.value) - 1n,
+              maxBytes.value == constants.defaultAggregatableFilteringIdMaxBytes
+                ? 'must be in the range [0, 255]. It exceeds the default max size of 1 byte. To increase, specify the aggregatable_filtering_id_max_bytes property.'
+                : undefined
+            )
+          })
```

Comment:
I think the logic here is complex enough that it merits being put in its own top-level function.